src/sankoff_handler.cpp

```diff
@@ -3,103 +3,158 @@
+        Assert(node->Children().size() == child_count,
+               "Error in SankoffHandler::RunSankoff: Tree should be bifurcating.");
+        // Which child partial is in right or left doesn't actually matter because they
+        // are summed when calculating q_partials. In DAG case, will want to ensure that
+        // left children partials are in p_partials_left
+        p_partials_left[pattern_idx].block<1, state_count>(node->Id(), 0) =
+            ParentPartial(TotalPPartial(p_partials_left[pattern_idx],
+                                        p_partials_right[pattern_idx],
+                                        node->Children()[0]->Id()));
+        p_partials_right[pattern_idx].block<1, state_count>(node->Id(), 0) =
+            ParentPartial(TotalPPartial(p_partials_left[pattern_idx],
+                                        p_partials_right[pattern_idx],
+                                        node->Children()[1]->Id()));
+      }
+    }
+  });
 
-    return {p_partials_left, p_partials_right, q_partials};
-} 
+  // generating q-partials
+  auto q_partials = BlankPartials(topology->Id() + 1);
+  topology->Preorder([&q_partials, &p_partials_left, &p_partials_right,
+                      this](const Node* node) {
+    if (!node->IsLeaf()) {
+      Assert(node->Children().size() == child_count,
+             "Error in SankoffHandler::RunSankoff: Tree should be bifurcating.");
+      for (size_t pattern_idx = 0; pattern_idx < site_pattern_.PatternCount();
+           pattern_idx++) {
+        auto partials_from_parent =
+            ParentPartial(q_partials[pattern_idx].block<1, state_count>(node->Id(), 0));
+        for (size_t child = 0; child < child_count; child++) {
+          size_t sister_idx = child == 0 ? 1 : 0;
+          auto partials_from_sister = ParentPartial(
+              TotalPPartial(p_partials_left[pattern_idx], p_partials_right[pattern_idx],
+                            node->Children()[sister_idx]->Id()));
+          q_partials[pattern_idx].block<1, state_count>(node->Children()[child]->Id(),
+                                                        0) =
+              partials_from_sister + partials_from_parent;
+        }
+      }
+    }
+  });
 
-double SankoffHandler::ParsimonyScore(std::tuple<std::vector<Eigen::Matrix<double, Eigen::Dynamic, SankoffHandler::state_count>>, std::vector<Eigen::Matrix<double, Eigen::Dynamic, SankoffHandler::state_count>>, std::vector<Eigen::Matrix<double, Eigen::Dynamic, SankoffHandler::state_count>>> all_partials, size_t node_id) {
-    // all_partials: {p_partials_left, p_partials_right, q_partials}
+  // Compress the p_partials_left, p_partials_right, and q_partials for one site
+  // together and use enum SankoffPartialType to separate the types.
+  auto concatenated_partials = BlankPartials(3 * (topology->Id() + 1));
+  for (size_t pattern = 0; pattern < site_pattern_.PatternCount(); pattern++) {
+    concatenated_partials[pattern] << p_partials_left[pattern],
```

Comment:
Whoa, is this Eigen magic?

Reply:
I did this to address the issue of passing the tuple around. Now for each site, there is a SankoffPartial where the first 3rd represents the left p-partials, etc.

Reply:
I get that, and it's awesome. I was just curious about the syntax here. Is `<<` an Eigen operator?

Reply:
`<<` is the Eigen comma initializer: https://eigen.tuxfamily.org/dox/group__TutorialAdvancedInitialization.html.